File: training-data/java/HashSetOpenAddressing.java
public class HashSetOpenAddressing {
    private static final Object TOMBSTONE = new Object();
    private Object[] slots = new Object[8];
    private int size, used;

    private int indexFor(Object key, int i) {
        int h = key.hashCode();
        h ^= (h >>> 16);
        return Math.floorMod(h + i, slots.length);
    }

    public boolean contains(Object key) {
        for (int i = 0; i < slots.length; i++) {
            Object s = slots[indexFor(key, i)];
            if (s == null) return false;
            if (s != TOMBSTONE && s.equals(key)) return true;
        }
        return false;
    }

    public boolean add(Object key) {
        if (contains(key)) return false;
        if ((used + 1) * 2 > slots.length) resize();
        for (int i = 0; ; i++) {
            int idx = indexFor(key, i);
            if (slots[idx] == null || slots[idx] == TOMBSTONE) {
                if (slots[idx] == null) used++;
                slots[idx] = key;
                size++;
                return true;
            }
        }
    }

    public boolean remove(Object key) {
        for (int i = 0; i < slots.length; i++) {
            int idx = indexFor(key, i);
            if (slots[idx] == null) return false;
            if (slots[idx] != TOMBSTONE && slots[idx].equals(key)) {
                slots[idx] = TOMBSTONE;
                size--;
                return true;
            }
        }
        return false;
    }

    private void resize() {
        Object[] old = slots;
        slots = new Object[old.length * 2];
        size = used = 0;
        for (Object o : old) if (o != null && o != TOMBSTONE) add(o);
    }

    public int size() { return size; }

    public static void main(String[] args) {
        HashSetOpenAddressing set = new HashSetOpenAddressing();
        for (String w : "the quick brown fox jumps over the lazy dog".split(" ")) set.add(w);
        set.remove("fox");
        System.out.println(set.size() + " " + set.contains("fox") + " " + set.contains("dog"));
    }
}

File: training-data/rust/hash_set_open_addressing.rs
use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};

enum Slot<K> {
    Empty,
    Tombstone,
    Full(K),
}

struct OpenSet<K> {
    slots: Vec<Slot<K>>,
    len: usize,
    used: usize,
}

impl<K: Hash + Eq> OpenSet<K> {
    fn new() -> Self {
        OpenSet { slots: (0..8).map(|_| Slot::Empty).collect(), len: 0, used: 0 }
    }

    fn index(&self, k: &K) -> usize {
        let mut h = DefaultHasher::new();
        k.hash(&mut h);
        (h.finish() as usize) % self.slots.len()
    }

    fn find(&self, k: &K) -> Option<usize> {
        let cap = self.slots.len();
        let mut i = self.index(k);
        for _ in 0..cap {
            match &self.slots[i] {
                Slot::Empty => return None,
                Slot::Full(x) if x == k => return Some(i),
                _ => i = (i + 1) % cap,
            }
        }
        None
    }

    fn insert(&mut self, k: K) -> bool {
        if self.find(&k).is_some() {
            return false;
        }
        if (self.used + 1) * 2 > self.slots.len() {
            let new_cap = self.slots.len() * 2;
            let old = std::mem::replace(&mut self.slots, (0..new_cap).map(|_| Slot::Empty).collect());
            self.len = 0;
            self.used = 0;
            for s in old {
                if let Slot::Full(x) = s {
                    self.insert(x);
                }
            }
        }
        let mut i = self.index(&k);
        while let Slot::Full(_) = self.slots[i] {
            i = (i + 1) % self.slots.len();
        }
        if let Slot::Empty = self.slots[i] {
            self.used += 1;
        }
        self.slots[i] = Slot::Full(k);
        self.len += 1;
        true
    }

    fn remove(&mut self, k: &K) -> bool {
        match self.find(k) {
            Some(i) => {
                self.slots[i] = Slot::Tombstone;
                self.len -= 1;
                true
            }
            None => false,
        }
    }

    fn contains(&self, k: &K) -> bool {
        self.find(k).is_some()
    }
}

fn main() {
    let mut s = OpenSet::new();
    for w in "the quick brown fox jumps over the lazy dog".split(' ') {
        s.insert(w);
    }
    s.remove(&"fox");
    println!("{} {} {}", s.len, s.contains(&"fox"), s.contains(&"dog"));
}
